Replace copy_memory with copy_from_slice + Fix ImageDecoder::load_rect

kngwyu opened the pull request:
I think it's make our code a bit simpler.

philipc commented on src/image.rs:
```diff
@@ -267,9 +267,8 @@ pub trait ImageDecoder: Sized {
-
-                ::copy_memory(from, to);
+                let start = i * width as usize;
+                buf[start..start + from.len()].copy_from_slice(from);
```
It looks like you're trying to fix a bug here at the same time, but it still looks wrong to me. I haven't tested, but I would expect it to be:
```rust
let from = &tmp[x as usize * bpp..][..width as usize * bpp];
let to = &mut buf[i * width as usize * bpp..][..width as usize * bpp];
to.copy_from_slice(from);
```

kngwyu replied:
Your code looks correct.
Thanks.

kngwyu commented:
Since `load_rect` had additional bugs other than @philipc commented, I added a test for it and fixed it.
In addition, how about changing its args from `(&mut self, x: u32, y: u32, length: u32, width: u32)` to `(&mut self, x: u32, y: u32, width: u32, length: u32)`, which is the same order as `GenericImage::sub_image`?
This is a breaking change, but I think not so many people will be affected by it, since this method is currently completely broken.

fintelia commented:
I'm currently working on more extreme breaking changes to this functionality in #819, so I'd advocate for holding off on changing the argument order here. Otherwise, I'm in favor of making the method slightly less buggy.

kngwyu commented:
Closed due to #819 merged
